core/src/oauth/providers/gong.rs

```diff
@@ -0,0 +1,161 @@
+use crate::{
+    oauth::{
+        connection::{
+            Connection, ConnectionProvider, FinalizeResult, Provider, ProviderError, RefreshResult,
+            PROVIDER_TIMEOUT_SECONDS,
+        },
+        providers::utils::execute_request,
+    },
+    utils,
+};
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use base64::{engine::general_purpose::STANDARD, Engine as _};
+use lazy_static::lazy_static;
+use serde_json::json;
+use std::env;
+
+lazy_static! {
+    static ref OAUTH_GONG_CLIENT_ID: String = env::var("OAUTH_GONG_CLIENT_ID").unwrap();
+    static ref OAUTH_GONG_CLIENT_SECRET: String = env::var("OAUTH_GONG_CLIENT_SECRET").unwrap();
+}
+
+pub struct GongConnectionProvider {}
+
+impl GongConnectionProvider {
+    pub fn new() -> Self {
+        GongConnectionProvider {}
+    }
+}
+
+#[async_trait]
+impl Provider for GongConnectionProvider {
+    fn id(&self) -> ConnectionProvider {
+        ConnectionProvider::Gong
+    }
+
+    async fn finalize(
+        &self,
+        _connection: &Connection,
+        code: &str,
+        redirect_uri: &str,
+    ) -> Result<FinalizeResult, ProviderError> {
+        let authorization = format!(
+            "Basic
+        {}",
+            STANDARD.encode(format!(
+                "{}:{}",
+                *OAUTH_GONG_CLIENT_ID, *OAUTH_GONG_CLIENT_SECRET
+            ))
+        );
+        let params = [
+            ("grant_type", "authorization_code"),
+            ("code", &code),
+            ("redirect_uri", &redirect_uri),
+        ];
+
+        let req = reqwest::Client::new()
+            .post("https://app.gong.io/oauth2/generate-customer-token")
+            .header("Content-Type", "application/json")
+            .header("Authorization", authorization)
+            .query(&params);
+
+        let raw_json = execute_request(ConnectionProvider::Gong, req)
+            .await
+            .map_err(|e| self.handle_provider_request_error(e))?;
+
+        let access_token = match raw_json["access_token"].as_str() {
+            Some(token) => token,
+            None => Err(anyhow!("Missing `access_token` in response from Gong"))?,
+        };
+
+        let expires_in = match raw_json.get("expires_in") {
+            Some(serde_json::Value::Number(n)) => match n.as_u64() {
+                Some(n) => n,
+                None => Err(anyhow!("Invalid `expires_in` in response from Gong"))?,
+            },
+            _ => Err(anyhow!("Missing `expires_in` in response from Gong"))?,
+        };
+
+        let refresh_token = match raw_json["refresh_token"].as_str() {
+            Some(token) => token,
+            None => Err(anyhow!("Missing `refresh_token` in response from Gong"))?,
+        };
+
+        Ok(FinalizeResult {
```

Comment:
From my understanding of their documentation, we also need to save `api_base_url_for_customer`. All queries should redirected to this url, do you confirm?

Reply:
Looks like it - missed this part 😓 
Will take another pass